src/coreclr/jit/codegenxarch.cpp

```diff
@@ -1106,6 +1106,21 @@ void CodeGen::genCodeForMul(GenTreeOp* treeNode)
 
             inst_RV_SH(INS_shl, size, targetReg, shiftAmount);
         }
+#ifdef TARGET_AMD64
+        else if (!requiresOverflowCheck && isUnsignedMultiply && rmOp->isUsedFromReg() && isPow2(imm + 1))
+        {
+            // Use shift for constant multiply when legal
+            uint64_t     zextImm     = static_cast<uint64_t>(static_cast<size_t>(imm + 1));
+            unsigned int shiftAmount = genLog2(zextImm);
+
+            // Copy reg src to dest register
+            inst_Mov(targetType, targetReg, rmOp->GetRegNum(), /* canSkip */ true);
+
+            inst_RV_SH(INS_shl, size, targetReg, shiftAmount);
+
+            GetEmitter()->emitIns_R_R(INS_sub, size, targetReg, rmOp->GetRegNum());
```

Comment:
The issue here is that `targetReg` could be the same as `rmOp->GetRegNum()` in which case this becomes `0`

You'd need to either request an extra temporary register in `LSRA` -or- do the transform from `mul` to `shl + sub` in `Lowering`.

I think doing the transform in `Lowering` will be overall better and allow other transforms to also lightup if applicable

Reply:
-- Noting that for `Lowering` if a similar transform was done on `Arm64` it would allow the existing `sub (shifted register)` to kick in for certain powers of 2. So its generally a better/more powerful approach to do it there IMO.

Reply:
It has been moved to lowering now.